pkg/operator/controller/externaldns/credentials-request.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externaldnscontroller
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	cco "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorv1alpha1 "github.com/openshift/external-dns-operator/api/v1alpha1"
+	controller "github.com/openshift/external-dns-operator/pkg/operator/controller"
+)
+
+// ensureExternalCredentialsRequest ensures that the externalDNS credential request exists.
+// Returns a boolean if the credential request exists, its current state if it exists
+// and an error if it cannot be created or updated.
+func (r *reconciler) ensureExternalCredentialsRequest(ctx context.Context, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *cco.CredentialsRequest, error) {
+	name := controller.ExternalDNSCredentialsRequestName(externalDNS)
+
+	desired := desiredCredentialsRequestName(name, externalDNS)
+
+	exists, current, err := r.currentExternalDNSCredentialsRequest(ctx, name)
+
+	if err != nil {
+		return false, nil, err
+	}
+
+	if !exists {
+		if err := r.createExternalDNSCredentialsRequest(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentExternalDNSCredentialsRequest(ctx, name)
+	}
+
+	return true, current, nil
+}
+
+func awsProviderConfig(codec *cco.ProviderCodec) (*runtime.RawExtension, error) {
+	awsProvSpec, err := codec.EncodeProviderSpec(
+		&cco.AWSProviderSpec{
+			TypeMeta: metav1.TypeMeta{
+				Kind: "AWSProviderSpec",
+			},
+			StatementEntries: []cco.StatementEntry{
+				{
+					Effect: "Allow",
+					Action: []string{
+						"elasticloadbalancing:DescribeLoadBalancers",
+						"route53:ListHostedZones",
+						"route53:ChangeResourceRecordSets",
+						"tag:GetResources"},
```

Comment:
1. Why do we need `elasticloadbalancing:DescribeLoadBalancers` ?
2. Why do we need `tag:GetResources`?
3. From [ExternalDNS AWS tutorial](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md#iam-policy) I see that `ListResourceRecordSets` action is also needed.
4. From the same tutorial I see that `ChangeResourceRecordSets` is limited to `arn:aws:route53:::hostedzone/*` resources, here we have it for all resources `*`, I'm wondering whether they are equivalent.